src/ua_securechannel.c

```diff
@@ -690,83 +520,269 @@ decryptAddChunk(UA_SecureChannel *channel, UA_MessageType messageType,
+
+    /* No complete mesage received */
+    if(finalType == UA_CHUNKTYPE_INTERMEDIATE)
+        return UA_STATUSCODE_GOOD;
+
+    /* Abort: Remove all chunks for this message */
+    if(finalType == UA_CHUNKTYPE_ABORT) {
+        while((chunk = SIMPLEQ_FIRST(&channel->completeChunks))) {
+            SIMPLEQ_REMOVE_HEAD(&channel->completeChunks, pointers);
+            SIMPLEQ_INSERT_TAIL(doneChunks, chunk, pointers);
+            if(chunk == finalChunk)
+                break;
+        }
+        return UA_STATUSCODE_GOOD;
+    }
+
+    /* A complete message has been received */
+    UA_assert(finalType == UA_CHUNKTYPE_FINAL);
+
+    /* Decrypt and verify the message chunks */
+    UA_Boolean first = true;
+    UA_UInt32 messageRequestId = 0;
+    UA_StatusCode res = UA_STATUSCODE_GOOD;
+    while((chunk = SIMPLEQ_FIRST(&channel->completeChunks))) {
+        /* Move the chunks for this message out of the channel queue */
+        SIMPLEQ_REMOVE_HEAD(&channel->completeChunks, pointers);
+        SIMPLEQ_INSERT_TAIL(doneChunks, chunk, pointers);
+
+        UA_UInt32 requestId;
+        res = decryptVerifySymmetricChunk(channel, sp, chunk, &requestId);
+        if(res != UA_STATUSCODE_GOOD)
+            return res;
+
+        /* Compare that the requestId is the same for all chunks */
+        if(first) {
+            messageRequestId = requestId;
+            first = false;
+        } else if(messageRequestId != requestId) {
+            return UA_STATUSCODE_BADSECURITYCHECKSFAILED;
+        }
+        if(chunk == finalChunk)
+            break;
+    }
+
+    return assembleProcessMessage(channel, doneChunks, chunksSize, messageRequestId,
+                                  messageType, application, callback);
+}
+
+static UA_StatusCode
+processCompleteChunks(UA_SecureChannel *channel, void *application,
+                      UA_ProcessMessageCallback callback) {
+    UA_StatusCode res = UA_STATUSCODE_GOOD;
+    UA_Chunk *chunk = SIMPLEQ_FIRST(&channel->completeChunks);
+    while(chunk)  {
+        UA_ChunkQueue doneChunks;
+        SIMPLEQ_INIT(&doneChunks);
+        switch(chunk->messageType) {
+        case UA_MESSAGETYPE_HEL:
+        case UA_MESSAGETYPE_ACK:
+        case UA_MESSAGETYPE_OPN:
+        case UA_MESSAGETYPE_ERR:
+            /* Chunks that are processed entirely by the application */
+            SIMPLEQ_REMOVE_HEAD(&channel->completeChunks, pointers);
+            SIMPLEQ_INSERT_TAIL(&doneChunks, chunk, pointers);
+            callback(application, channel, chunk->messageType, 0, &chunk->bytes);
+            break;
+        default: /* MSG and CLO */
+            res = processSymmetricChunks(channel, &doneChunks, application, callback);
+            break;
+        }
+
+        deleteChunks(&doneChunks); 
+        if(res != UA_STATUSCODE_GOOD)
+            break;
+
+        /* Could we process a full message? Done? */
+        UA_Chunk *next = SIMPLEQ_FIRST(&channel->completeChunks);
+        if(chunk == next)
```

Comment:
I think i understand what we are doing here, but clang analyzer thinks this condition is always true, which is not the case.
Might indicate a problem that i don't see or too complex code. (And maybe a bug in clang analyzer of course)